todomvc/assets/js/app.js

```diff
@@ -0,0 +1,440 @@
+      }
+      
+      this.render()
+    } catch (error) {
+      console.error('Failed to update todo:', error)
+    }
+  }
+
+  async deleteTodo(id) {
+    if (!this.contractID) return
+    
+    try {
+      // Delete directly from KV store
+      await sbp('chelonia/kv/delete', this.contractID, `todo:${id}`)
+      
+      // Update local state
+      delete this.todos[id]
+      
+      this.render()
+    } catch (error) {
+      console.error('Failed to delete todo:', error)
+    }
+  }
+
+  async toggleAll(completed) {
+    if (!this.contractID) return
+    
+    try {
+      // Update all todos directly in KV store
+      for (const [id, todo] of Object.entries(this.todos)) {
+        const updatedTodo = { ...todo, completed, updatedDate: new Date().toISOString() }
+        await sbp('chelonia/kv/set', this.contractID, `todo:${id}`, JSON.stringify(updatedTodo))
+        this.todos[id] = updatedTodo
+      }
+      
+      this.render()
+    } catch (error) {
+      console.error('Failed to toggle all todos:', error)
+    }
+  }
+
+  async clearCompleted() {
+    if (!this.contractID) return
+    
+    try {
+      // Delete completed todos directly from KV store
+      const completedIds = Object.keys(this.todos).filter(id => this.todos[id].completed)
+      
+      for (const id of completedIds) {
+        await sbp('chelonia/kv/delete', this.contractID, `todo:${id}`)
+        delete this.todos[id]
+      }
+      
+      this.render()
+    } catch (error) {
+      console.error('Failed to clear completed todos:', error)
+    }
+  }
+
+  setFilter(filter) {
+    this.currentFilter = filter
+    
+    // Update filter links
+    this.filterLinks.forEach(link => {
+      link.classList.remove('selected')
+      if (link.getAttribute('href') === `#/${filter}` || (filter === 'all' && link.getAttribute('href') === '#/')) {
+        link.classList.add('selected')
+      }
+    })
+    
+    this.render()
+  }
+
+  getFilteredTodos() {
+    const todos = Object.values(this.todos)
+    
+    switch (this.currentFilter) {
+      case 'active':
+        return todos.filter(todo => !todo.completed)
+      case 'completed':
+        return todos.filter(todo => todo.completed)
+      default:
+        return todos
+    }
+  }
+
+  render() {
+    const todos = Object.values(this.todos)
+    const filteredTodos = this.getFilteredTodos()
+    const activeTodos = todos.filter(todo => !todo.completed)
+    const completedTodos = todos.filter(todo => todo.completed)
+    
+    // Show/hide main and footer sections
+    const hasTodos = todos.length > 0
+    this.mainSection.style.display = hasTodos ? 'block' : 'none'
+    this.footerSection.style.display = hasTodos ? 'block' : 'none'
+    
+    // Update toggle all checkbox
+    this.toggleAllCheckbox.checked = activeTodos.length === 0 && todos.length > 0
+    
+    // Render todo list
+    this.todoList.innerHTML = filteredTodos.map(todo => this.renderTodoItem(todo)).join('')
```

Comment:
I realise this is a demo app, but still `.innerHTML` is a code smell. There are at least two ways you can address this. The 'classical' way, i.e., building DOM elements using `document.createElement` and the 'modern' way using the `<template>` element.